File: text-mate-core/src/main/java/io/github/rosemoe/sora/textmate/core/theme/ThemeTrieElementRule.java
/**
 * Copyright (c) 2015-2017 Angelo ZERR.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 * <p>
 * SPDX-License-Identifier: EPL-2.0
 * <p>
 * Contributors:
 * Angelo Zerr <angelo.zerr@gmail.com> - initial API and implementation
 */
package io.github.rosemoe.sora.textmate.core.theme;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

public class ThemeTrieElementRule {

    // _themeTrieElementRuleBrand: void;

    public final List<String> parentScopes;
    public int scopeDepth;
    public int fontStyle;
    public int foreground;
    public int background;

    public ThemeTrieElementRule(int scopeDepth, List<String> parentScopes, int fontStyle, int foreground,
                                int background) {
        this.scopeDepth = scopeDepth;
        this.parentScopes = parentScopes;
        this.fontStyle = fontStyle;
        this.foreground = foreground;
        this.background = background;
    }

    public static List<ThemeTrieElementRule> cloneArr(List<ThemeTrieElementRule> arr) {
        List<ThemeTrieElementRule> r = new ArrayList<>();
        for (int i = 0, len = arr.size(); i < len; i++) {
            r.add(arr.get(i).clone());
        }
        return r;
    }

    @Override
    public ThemeTrieElementRule clone() {
        return new ThemeTrieElementRule(this.scopeDepth, this.parentScopes, this.fontStyle, this.foreground,
                this.background);
    }

    public void acceptOverwrite(int scopeDepth, int fontStyle, int foreground, int background) {
        if (this.scopeDepth > scopeDepth) {
            // TODO!!!
            // console.log('how did this happen?');
        } else {
            this.scopeDepth = scopeDepth;
        }
        // console.log('TODO -> my depth: ' + this.scopeDepth + ', overwriting
        // depth: ' + scopeDepth);
        if (fontStyle != FontStyle.NotSet) {
            this.fontStyle = fontStyle;
        }
        if (foreground != 0) {
            this.foreground = foreground;
        }
        if (background != 0) {
            this.background = background;
        }
    }

    @Override
    public int hashCode() {
        return Objects.hash(background, fontStyle, foreground, parentScopes, scopeDepth);
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj) {
            return true;
        }
        if (obj == null) {
            return false;
        }
        if (getClass() != obj.getClass()) {
            return false;
        }
        ThemeTrieElementRule other = (ThemeTrieElementRule) obj;
        return background == other.background && fontStyle == other.fontStyle && foreground == other.foreground &&
                Objects.equals(parentScopes, other.parentScopes) && scopeDepth == other.scopeDepth;
    }


}

File: language-monarch/src/main/java/io/github/rosemoe/sora/langs/monarch/theme/ThemeTrieElementRule.kt
/*******************************************************************************
 * Copyright (c) 2015-2017 Angelo ZERR.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 * <p>
 * SPDX-License-Identifier: EPL-2.0
 * <p>
 * Contributors:
 * Angelo Zerr <angelo.zerr@gmail.com> - initial API and implementation
 * Sebastian Thomschke (Vegard IT GmbH) - add previousLineText support
 *
 ******************************************************************************/

package io.github.rosemoe.sora.langs.monarch.theme

import io.github.dingyi222666.monarch.types.ColorId
import io.github.dingyi222666.monarch.types.FontStyle
import io.github.dingyi222666.monarch.types.MetadataConsts

class ThemeTrieElementRule(
    private var fontStyle: Int,
    private var foreground: Int,
    private var background: Int
) {
    var metadata: Int = (fontStyle shl MetadataConsts.FONT_STYLE_OFFSET) or
            (foreground shl MetadataConsts.FOREGROUND_OFFSET) or
            (background shl MetadataConsts.BACKGROUND_OFFSET)

    fun clone(): ThemeTrieElementRule = ThemeTrieElementRule(fontStyle, foreground, background)

    fun acceptOverwrite(fontStyle: Int, foreground: Int, background: Int) {
        if (fontStyle != FontStyle.NotSet) {
            this.fontStyle = fontStyle
        }
        if (foreground != ColorId.None) {
            this.foreground = foreground
        }
        if (background != ColorId.None) {
            this.background = background
        }
        metadata = (this.fontStyle shl MetadataConsts.FONT_STYLE_OFFSET) or
                (this.foreground shl MetadataConsts.FOREGROUND_OFFSET) or
                (this.background shl MetadataConsts.BACKGROUND_OFFSET)
    }

    override fun toString(): String {
        return "ThemeTrieElementRule(fontStyle=$fontStyle, foreground=$foreground, background=$background, metadata=$metadata)"
    }


}
